### PR TITLE
[WIP] [Locale] Intl C4R

### PR DESCRIPTION
**Pull request by menusal**
# Description

Shortcut: [145211](https://app.shortcut.com/cartoteam/story/145211/add-multi-language)

Use multi-language in projects

## Type of change

(choose one and remove the others)

- Feature

# Acceptance

Please describe how to validate the feature or fix

1. go to X
2. test Y
3. assert Z

If feature deals with theme / UI or internal elements used also in CARTO 3, please also add a note on how to do acceptance on that part.

# Basic checklist

- Good PR name
- Shortcut link
- Changelog entry
- Just one issue per PR
- GitHub labels
- Proper status & reviewers
- Tests
- Documentation

### CODE REVIEW

File: packages/react-ui/src/widgets/TimeSeriesWidgetUI/TimeSeriesWidgetUI.js [hunk 1]
```diff
@@ -257,21 +260,26 @@ function TimeSeriesWidgetUIContent({
     }
 
     const formatter = FORMAT_DATE_BY_STEP_SIZE[stepSize];
+    const weeOfMessage = intl.formatMessage({ id: 'c4r.widgets.timeSeries.weekOf' });
```
Comment (zbigg) [inline packages/react-ui/src/widgets/TimeSeriesWidgetUI/TimeSeriesWidgetUI.js:263]: minor weekOfMesssage

File: packages/react-ui/src/hooks/useImperativeIntl.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import { useContext, useMemo } from 'react';
+import { match } from '@formatjs/intl-localematcher';
+
+export const DEFAULT_LOCALE = 'en-US';
+const cache = createIntlCache();
+
+// TODO: Pending to get external intl provider
+
+export default function useImperativeIntl() {
+  const extenal = useIntl();
```
Comment (VictorVelarde) [inline packages/react-ui/src/hooks/useImperativeIntl.js:12]: typo 'external'?

File: packages/react-ui/src/localization/id-ID.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
+const appMessages = {
```
Comment (VictorVelarde) [inline packages/react-ui/src/localization/id-ID.js:1]: If this file is not required (just for demo purpose), I would recommend to delete it

File: packages/react-ui/src/widgets/RangeWidgetUI/RangeWidgetUI.js [hunk 4]
```diff
@@ -74,20 +75,25 @@ const SliderLimit = styled(Slider)(({ theme: { palette, spacing } }) => ({
 
 /**
  * Renders a <RangeWidget /> component
+ *
+ * <!--
```
Comment (VictorVelarde) [inline packages/react-ui/src/widgets/RangeWidgetUI/RangeWidgetUI.js:79]: Why adding this?

File: packages/react-ui/src/hooks/useImperativeIntl.js (hunk 2 above)
Comment (zbigg) [reply packages/react-ui/src/hooks/useImperativeIntl.js:12]: it shouldn't be here actually:

```
const extenal = useIntl();
const externalIntl = useContext(IntlContext);
```

take same object, but first asserts (and throws!) when context is null, so this is undesirable, we would like to leave as much backwards compatibility as needed and make ReactIntlProvider optional ...

second, just gets value - if absent, it falls back to null and we can use our defaults

**Comment by VictorVelarde** [comment]
Hi @aaranadev this is almost there. 

If you can please have a general look / do some basic cleanup I think we'll be ready for an alpha. With the alpha, we will make a PR on cloud-native to check it works and we will appreciate a very small PR in carto-react-template to validate it works as expected

File: packages/react-ui/src/widgets/TimeSeriesWidgetUI/TimeSeriesWidgetUI.js (hunk 1 above)
Comment (menusal) [reply packages/react-ui/src/widgets/TimeSeriesWidgetUI/TimeSeriesWidgetUI.js:263]: fixed

File: packages/react-ui/src/widgets/legend/LegendWrapper.js [hunk 5]
```diff
@@ -182,7 +187,14 @@ function Header({
         </Tooltip>
       )}
       {switchable && (
-        <Tooltip title={(visible ? 'Hide' : 'Show') + ' layer'}>
```
[the remark this answers is not in the document]
Comment (menusal) [reply packages/react-ui/src/widgets/legend/LegendWrapper.js:185]: fixed

**Comment by menusal** [comment]
> Hi @aaranadev this is almost there.
> 
> If you can please have a general look / do some basic cleanup I think we'll be ready for an alpha. With the alpha, we will make a PR on cloud-native to check it works and we will appreciate a very small PR in carto-react-template to validate it works as expected

@VictorVelarde @zbigg @aaranadev 
I have resumed PR. Please check the latest commits.
Next step before merging the PR. Try using the Provider in the template and check that the language changes.

It's unclear to me that it works without exposing the context using the imperative API. 🤔

File: packages/react-ui/src/hooks/useImperativeIntl.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+export const DEFAULT_LOCALE = 'en';
+const cache = createIntlCache();
+
+export default function useImperativeIntl(intlConfig) {
+  const locale = intlConfig?.locale || DEFAULT_LOCALE;
+  const messagesLocale = findMatchingMessagesLocale(locale);
+  const intMessages = intlConfig?.messages
+    ? flattenMessages(intlConfig?.messages)
+    : messages[messagesLocale];
```
Comment (zbigg) [inline packages/react-ui/src/hooks/useImperativeIntl.js:15]: i think all lines above should be in memo together with `createIntl`

mainly `findMatchingMessagesLocale` and `flattenMessages` are not trivial and everything depends _only_ on `intlConfig`

File: packages/react-widgets/src/widgets/TimeSeriesWidget.js [hunk 7]
```diff
@@ -81,6 +81,7 @@ const EMPTY_ARRAY = [];
  * @param  {string[]=} [props.palette] - Optional palette
  * @param  {object} [props.wrapperProps] - Extra props to pass to [WrapperWidgetUI](https://storybook-react.carto.com/?path=/docs/widgets-wrapperwidgetui--default).
  * @param  {object} [props.noDataAlertProps] - Extra props to pass to [NoDataAlert]().
+ * @param  {object} [props.intlConfig] - Object with intl configuration. If not provided, default messages will be used.
```
Comment (zbigg) [inline packages/react-widgets/src/widgets/TimeSeriesWidget.js:84]: please use `{object=}` to indicate that this is [optional parameter](https://stackoverflow.com/a/29441515)

(not necessarily here, as there is proper .d.ts, but mainly for other widgets and places when you add optional param

File: packages/react-ui/src/hooks/useImperativeIntl.js (hunk 6 above)
Comment (menusal) [reply packages/react-ui/src/hooks/useImperativeIntl.js:15]: Done

**Pull request closed without merging**